klippa/src/hvar.rs

```diff
@@ -0,0 +1,640 @@
+impl Subset for Hvar<'_> {
+    fn subset(
+        &self,
+        plan: &Plan,
+        _font: &FontRef,
+        s: &mut Serializer,
+        _builder: &mut FontBuilder,
+    ) -> Result<(), SubsetError> {
+        let var_store = self
+            .item_variation_store()
+            .map_err(|_| SubsetError::SubsetTableError(Hvar::TAG))?;
+
+        let index_maps = self
+            .listup_index_maps()
+            .map_err(|_| SubsetError::SubsetTableError(Hvar::TAG))?;
+
+        let hvar_subset_plan = HvarVvarSubsetPlan::new(plan, &var_store, &index_maps)
+            .map_err(|_| SubsetError::SubsetTableError(Hvar::TAG))?;
+
+        s.embed(self.version())
+            .map_err(|_| SubsetError::SubsetTableError(Hvar::TAG))?;
+
+        let var_store_offset_pos = s
+            .embed(0_u32)
+            .map_err(|_| SubsetError::SubsetTableError(Hvar::TAG))?;
+
+        Offset32::serialize_subset(
+            &var_store,
+            s,
+            plan,
+            hvar_subset_plan.inner_maps(),
+            var_store_offset_pos,
+        )
+        .map_err(|_| SubsetError::SubsetTableError(Hvar::TAG))?;
+
+        serialize_index_maps(
+            s,
+            plan,
+            &index_maps,
+            hvar_subset_plan.index_map_subset_plans(),
+        )
+        .map_err(|_| SubsetError::SubsetTableError(Hvar::TAG))
+    }
+}
+
+pub(crate) trait ListupIndexMaps {
+    fn listup_index_maps(&self) -> Result<Vec<Option<DeltaSetIndexMap>>, ReadError>;
+}
+
+impl ListupIndexMaps for Hvar<'_> {
+    fn listup_index_maps(&self) -> Result<Vec<Option<DeltaSetIndexMap>>, ReadError> {
+        let out = vec![
+            self.advance_width_mapping().transpose()?,
+            self.lsb_mapping().transpose()?,
+            self.rsb_mapping().transpose()?,
+        ];
+        Ok(out)
+    }
+}
+
+pub(crate) fn serialize_index_maps(
+    s: &mut Serializer,
+    plan: &Plan,
+    index_maps: &[Option<DeltaSetIndexMap>],
+    index_map_plans: &[IndexMapSubsetPlan],
+) -> Result<(), SerializeErrorFlags> {
+    if index_maps.len() != index_map_plans.len() {
+        return Err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+    }
+
+    for (index_map, index_map_subset_plan) in index_maps.iter().zip(index_map_plans) {
+        let offset_pos = s.embed(0_u32)?;
+        if index_map_subset_plan.is_identity() {
+            continue;
+        }
+
+        Offset32::serialize_subset(
+            index_map.as_ref().unwrap(),
+            s,
+            plan,
+            &index_map_subset_plan.to_serialize_plan(),
+            offset_pos,
+        )?;
+    }
+    Ok(())
+}
+
+#[derive(Default)]
+pub(crate) struct IndexMapSubsetPlan {
+    map_count: u16,
+    max_inners: Vec<u16>,
+    outer_bit_count: u8,
+    inner_bit_count: u8,
+    output_map: FnvHashMap<GlyphId, u32>,
+}
+
+impl IndexMapSubsetPlan {
+    fn init(
+        index_map: &Option<DeltaSetIndexMap>,
```

Comment:
you rarely want `&Option<T>`; `Option<&T>` is more flexible, and it's easier to convert from `&Option<T>` with `.as_ref()`.